File: scripts/javascript/bootstrap.js
const { CrustSDK } = require('./sdk.js');
const fs = require('fs');
const vm = require('vm');

async function bootstrap() {
    const sdk = new CrustSDK();
    const event = JSON.parse(Buffer.from(process.env.USER_EVENT, 'base64').toString());
    const userCode = Buffer.from(process.env.USER_CODE, 'base64').toString('utf-8');

    const context = {
        // Essential globals
        global,
        process,
        console,
        Buffer,
        setTimeout,
        setInterval,
        setImmediate,
        clearTimeout,

        // Modules and Exports
        require: require, // Give them the REAL require
        exports: {},
        module: { exports: {} },
    };

    vm.createContext(context);

    try {
        vm.runInContext(userCode, context);

        if (typeof context.exports.handler === 'function') {
            const result = await context.exports.handler(sdk, event);
            await safeOutput(result);
        }
    } catch (err) {
        await safeOutput({ error: err.message });
        process.exit(1);
    }
}

async function safeOutput(result) {
    const outputPath = '/output/result.json';
    try {
        fs.writeFileSync(outputPath, JSON.stringify(result));
        console.log('Output written successfully at ', outputPath);
    } catch (err) {
        console.error('Failed to write output:', err);
    }
}

bootstrap();

File: scripts/python/bootstrap.py
# This file is used to bootstrap the project by running necessary setup tasks, such as starting the worker and ensuring dependencies are in place.

import os, base64, json
from sdk import CrustSDK

sdk = CrustSDK()
event = json.loads(base64.b64decode(os.getenv('USER_EVENT')))
user_code = base64.b64decode(os.getenv('USER_CODE')).decode('utf-8')

user_module = {}
exec(user_code, user_module)

def safe_output(result):
    RW_OUTPUT_PATH = '/output/result.json';
    with open(RW_OUTPUT_PATH, 'w') as f:
        json.dump(result, f)

if 'handler' in user_module:
    try:
        result = user_module['handler'](sdk, event)
        safe_output(result)
    except Exception as e:
        safe_output({"error": str(e), "kind": e.__class__.__name__})
        exit(1)
else:
    safe_output({"error": "No handler function found in user code."})
    exit(1)